File: julia/src/benchmarks/Mandelbrot.jl
mutable struct Mandelbrot <: AbstractBenchmark
    w::Int64
    h::Int64
    io::IOBuffer

    function Mandelbrot()
        w = Helper.config_i64("Mandelbrot", "w")
        h = Helper.config_i64("Mandelbrot", "h")
        new(w, h, IOBuffer())
    end
end

name(b::Mandelbrot)::String = "Mandelbrot"

function run(b::Mandelbrot, iteration_id::Int64)
    ITER = 50
    LIMIT = 2.0
    LIMIT_SQ = LIMIT * LIMIT

    io = b.io

    write(io, "P4\n$(b.w) $(b.h)\n")

    bit_num = 0
    byte_acc = UInt8(0)

    h = b.h
    w = b.w

    for y in 0:h-1
        ci = (2.0 * y / h - 1.0)

        for x in 0:w-1
            cr = (2.0 * x / w - 1.5)

            zr = zi = tr = ti = 0.0

            i = 0
            while i < ITER && (tr + ti) <= LIMIT_SQ
                zi = 2.0 * zr * zi + ci
                zr = tr - ti + cr
                tr = zr * zr
                ti = zi * zi
                i += 1
            end

            byte_acc <<= 1
            if tr + ti <= LIMIT_SQ
                byte_acc |= 0x01
            end

            bit_num += 1

            if bit_num == 8
                write(io, byte_acc)
                byte_acc = 0x00
                bit_num = 0
            elseif x == w - 1
                byte_acc <<= (8 - (w % 8))
                write(io, byte_acc)
                byte_acc = 0x00
                bit_num = 0
            end
        end
    end
end

function checksum(b::Mandelbrot)::UInt32
    return Helper.checksum(take!(b.io))
end

File: nim/src/benchmarks/mandelbrot.nim
import std/[math, strutils]
import ../benchmark
import ../helper

type
  Mandelbrot* = ref object of Benchmark
    w, h: int64
    resultBin: seq[byte]

proc newMandelbrot(): Benchmark =
  Mandelbrot()

method name(self: Mandelbrot): string = "CLBG::Mandelbrot"

method prepare(self: Mandelbrot) =
  self.w = self.config_val("w")
  self.h = self.config_val("h")
  self.resultBin = @[]

method run(self: Mandelbrot, iteration_id: int) =
  const ITER = 50
  const LIMIT = 2.0

  let header = "P4\n" & $self.w & " " & $self.h & "\n"
  for c in header:
    self.resultBin.add(byte(c))

  var bitNum = 0
  var byteAcc: byte = 0

  for y in 0..<self.h.int:
    for x in 0..<self.w.int:
      let tmpX = float(x)
      let tmpY = float(y)
      let tmpW = float(self.w)
      let tmpH = float(self.h)

      let cr = 2.0 * tmpX / tmpW - 1.5
      let ci = 2.0 * tmpY / tmpH - 1.0

      var zr = 0.0
      var zi = 0.0
      var tr = 0.0
      var ti = 0.0

      var i = 0
      while i < ITER and tr + ti <= LIMIT * LIMIT:
        zi = 2.0 * zr * zi + ci
        zr = tr - ti + cr
        tr = zr * zr
        ti = zi * zi
        inc i

      byteAcc = byteAcc shl 1
      if tr + ti <= LIMIT * LIMIT:
        byteAcc = byteAcc or 0x01
      inc bitNum

      if bitNum == 8:
        self.resultBin.add(byteAcc)
        byteAcc = 0
        bitNum = 0
      elif x == self.w.int - 1:
        let shiftAmount = 8 - (self.w.int mod 8)
        byteAcc = byteAcc shl shiftAmount
        self.resultBin.add(byteAcc)
        byteAcc = 0
        bitNum = 0

method checksum(self: Mandelbrot): uint32 =
  checksum(self.resultBin)

registerBenchmark("CLBG::Mandelbrot", newMandelbrot)
